### PR TITLE
Add Combined Bremsstrahlung Model

### PR DESCRIPTION
**Pull request by whokion**
This MR includes
- a combined bremsstrahlung model (CombinedBremModel) with SeltzerBergerModel and RelativisticBremModel and an associated test
- a bug fix for the angular sampling in SeltzerBergerInteractor
- a new header for derivative constants for physics models, PhysicsConstants
- updates for SeltzerBerger.test and RelativisticBrem.test due to changes in material properties, cutoff and etc.
A follow-up MR for adding RelativisticBremModel and CombinedBremModel into BremsstrahlungProcess and related changes for the demo-loop are also ready and will be opened once this MR is merged.

### CODE REVIEW

**Review by sethrj — CHANGES_REQUESTED** [review]
The way you've combined the `Model` classes is nice. The interactor though needs a little work, Currently the `update_state` is effectively duplicated across the three brems interactors. I think you should make a small distribution-like helper class that has the call signature `Interaction operator()(Engine& rng);`. Then you should be able to return that value from all three Interactor classes. Since the relativistic exiting energy sampler is duplicated, that too should be a small helper class.

**Comment by whokion** [comment]
> The way you've combined the `Model` classes is nice. The interactor though needs a little work, Currently the `update_state` is effectively duplicated across the three brems interactors. I think you should make a small distribution-like helper class that has the call signature `Interaction operator()(Engine& rng);`. Then you should be able to return that value from all three Interactor classes. Since the relativistic exiting energy sampler is duplicated, that too should be a small helper class.

@sethrj Agreed!  I was actually trying to add utility functions for sampling energy and angular distributions that could commonly used in three interactors, which makes code maintenance consistently, but decided to refactor codes later as s the first implementation looked ugly passing too many arguments to them - introducing the electron_mass_mev (which is not used in the current code) was a part of avoiding passing a data to function from them.  I also thought about a helper class, but hesitate to add too many helper classes for this MR. The original simple (but much cleaner) approach to use operators of SB and relativistic interactors directly did not worked as it crashes somewhere in secondary memory allocations during testing codes which I could not understand fully.  Also such approach might degrade the computing performance as two interactors are created in every call for the combined operator.  Anyway, I will introduce helper classes and update the code accordingly. Thanks for your comments - I did not mean to interrupt your vacation, but put the branch for the earlier test.

**Comment by sethrj** [comment]
> @sethrj Agreed! I was actually trying to add utility functions for sampling energy and angular distributions that could commonly used in three interactors, which makes code maintenance consistently, but decided to refactor codes later as s the first implementation looked ugly passing too many arguments to them - introducing the electron_mass_mev (which is not used in the current code) was a part of avoiding passing a data to function from them. I also thought about a helper class, but hesitate to add too many helper classes for this MR.

I'm always OK with adding more helper classes 😄especially if they factor out duplicate code.

> The original simple (but much cleaner) approach to use operators of SB and relativistic interactors directly did not worked as it crashes somewhere in secondary memory allocations during testing codes which I could not understand fully. Also such approach might degrade the computing performance as two interactors are created in every call for the combined operator.

The crash is probably a matter of keeping a reference to a temporary: a lot of our helper classes do that and it can make them somewhat fragile.

> Anyway, I will introduce helper classes and update the code accordingly. Thanks for your comments - I did not mean to interrupt your vacation, but put the branch for the earlier test.

No interruption at all! I'm doing a little bit of work here and there, so it's not completely vacation 😉

File: src/physics/em/detail/PhysicsConstants.hh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * Derived constants    | Unit                  | Notes
+ * -------------------- | --------------------- | ------------
+ * electron_mass_c2     | g * (m/s)^2           |
+ * electron_mass_mev    | g * (m/s)^2/MeV       | Geant4 electron mass
+ * migdal_constant      | cm^3                  | Bremsstrahlung
+ * lpm_constant         | MeV/cm                | Relativistic Bremsstrahlung
+ */
+
+using namespace constants;
+using namespace units;
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/physics/em/detail/PhysicsConstants.hh]: Replaced constants by constant functions for now.

File: src/physics/em/detail/PhysicsConstants.hh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+//!@{
+//! Derivative constants with units
+constexpr real_type electron_mass_c2  = electron_mass * c_light * c_light;
+constexpr real_type electron_mass_mev = electron_mass_c2
+                                        / (1e6 * e_electron * volt);
+constexpr real_type migdal_constant = 4 * pi * r_electron * lambdabar_electron
+                                      * lambdabar_electron;
+constexpr real_type lpm_constant = alpha_fine_structure * electron_mass_c2
+                                   * electron_mass_c2
+                                   / (2 * h_planck * c_light);
```
[the remark this answers is not in the document]
Comment (whokion) [reply src/physics/em/detail/PhysicsConstants.hh]: Dropped electron_mass_mev as is is not used in the current code.  I think that defining this somewhere in physics will remove a lot of duplicated shared data (i.e., in various model data).

**Comment by sethrj** [comment]
Yeah the CI error is "no more space left on device". Reviewing updates now.

File: src/physics/em/detail/BremFinalStateHelper.hh [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                                               const Real3&      inc_direction,
+                                               const Momentum&   inc_momentum,
+                                               const Mass&       inc_mass,
+                                               const ParticleId& gamma_id);
+
+    // Update the final state for the given RNG and the photon energy
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine&      rng,
+                                                 const Energy gamma_energy,
+                                                 Secondary*   secondaries);
```
Comment (sethrj) [inline src/physics/em/detail/BremFinalStateHelper.hh:44]: These last two arguments should be part of the constructor as well: it's "given the incident particle data and the exiting gamma energy, what is one realization of the exiting interaction" (which will all have the same energy change but different directions)

Comment (sethrj) [reply src/physics/em/detail/BremFinalStateHelper.hh:44]: To make this easier as a really simple helper function you could also just write this whole thing as a simple struct with public data members, then construct and call it like:
```c++
BremInteractionSampler sample_interaction{inc_energy, inc_direction, inc_momentum, inc_mass, gamma_energy, gamma_id};
return sample_interaction(rng);
```
Or you could get fancier and keep it as a class but create sub-distributions as class members, precalculating the exit energy and having the various samplers as member data. The only *real* advantage to doing it that way, though, would be to reduce the size of templated code if we had multiple RNG instantiations, which in practice we won't have. You could also argue that it improves the structure by separating the "setup" based on knowns from the "sampling."

File: src/physics/em/detail/BremFinalStateHelper.i.hh [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Update parent particle direction
+    for (unsigned int i : range(3))
+    {
+        real_type inc_momentum_i   = inc_momentum_.value() * inc_direction_[i];
+        real_type gamma_momentum_i = result.secondaries[0].energy.value()
+                                     * result.secondaries[0].direction[i];
+        result.direction[i] = inc_momentum_i - gamma_momentum_i;
+    }
+    normalize_direction(&result.direction);
```
Comment (sethrj) [inline src/physics/em/detail/BremFinalStateHelper.i.hh:67]: Just a note (not part of this PR) but we do secondary/primary momentum balance in many of our interactors so should probably add a helper function for these 8 lines.

File: src/physics/em/detail/CombinedBremInteractor.hh [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+#include "SBEnergySampler.hh"
+#include "BremFinalStateHelper.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Brief class description.
```
Comment (sethrj) [inline src/physics/em/detail/CombinedBremInteractor.hh:36]: Comment needs updating.

File: src/physics/em/detail/CombinedBremInteractor.i.hh [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        // Failed to allocate space for the secondary
+        return Interaction::from_failure();
+    }
+
+    // Sample the bremsstrahlung photon energy
+    Energy gamma_energy = (is_relativistic_) ? rb_energy_sampler_(rng)
+                                             : sb_energy_sampler_(rng);
+
+    // Update kinematics of the final state and return this interaction
+    return final_state_interaction_(rng, gamma_energy, secondaries);
```
Comment (sethrj) [inline src/physics/em/detail/CombinedBremInteractor.i.hh:89]: Aha, my comment on changing the signature would require the final state interaction to become "local" data rather than "state" data. So it could go either way.

File: src/physics/em/detail/PhysicsConstants.hh [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Physical constants which are derived from fundamental constants.
+ *
+ * Derived constants    | Unit                  | Notes
+ * -------------------- | --------------------- | ------------
+ * electron_mass_c2()   | g * (m/s)^2           |
```
Comment (sethrj) [inline src/physics/em/detail/PhysicsConstants.hh:24]:
```suggestion
 * electron_mass_c2()   | g * (cm/s)^2           |
```
based on the unit system

File: src/physics/em/detail/PhysicsConstants.hh [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+CELER_CONSTEXPR_FUNCTION real_type migdal_constant()
+{
+    using namespace constants;
+    using namespace units;
+
+    return 4 * pi * r_electron * lambdabar_electron * lambdabar_electron;
+}
+
+CELER_CONSTEXPR_FUNCTION real_type lpm_constant()
```
Comment (sethrj) [inline src/physics/em/detail/PhysicsConstants.hh:46]: This return type is inconsistent with the units in your documentation I think? Should this be returning a quantity in `MevEnergy`? And if that's confusing (since this is energy per "native" length) you could define a new quantity in this file:
```
using MevPerCm  = Quantity<UnitDivide<units::Mev, units::NativeUnit>>;
```

File: src/physics/em/detail/RBEnergySampler.i.hh [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+template<class Engine>
+CELER_FUNCTION auto RBEnergySampler::operator()(Engine& rng) -> Energy
+{
+    // Min and max kinetic energy limits for sampling the secondary photon
+    Energy tmin = min(gamma_cutoff_, inc_energy_);
+    Energy tmax = min(shared_.high_energy_limit(), inc_energy_);
+
+    real_type density_corr = dxsec_.density_correction();
+    real_type xmin         = std::log(ipow<2>(tmin.value()) + density_corr);
+    real_type xrange = std::log(ipow<2>(tmax.value()) + density_corr) - xmin;
```
Comment (sethrj) [inline src/physics/em/detail/RBEnergySampler.i.hh:48]:
```suggestion
    ReciprocalSampler sample_ee(ipow<2>(min_gamma_energy) + density_corr,
                             ipow<2>(inc_energy) + density_corr)
```
and below replace 
```c++
std::exp(xmin + generate_canonical(rng) * xrange)
```
with
```
sample_ee(rng);
```

This is basically the same (aside from the XS calculation) as is done in `SBEnergyDistHelper`, so maybe we can further reduce code duplication by templating that on the cross section code.

File: src/physics/em/detail/RBEnergySampler.i.hh [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+    real_type gamma_energy{0};
+    real_type dsigma{0};
+
+    do
+    {
+        gamma_energy = std::sqrt(max(
+            real_type(0),
+            std::exp(xmin + generate_canonical(rng) * xrange) - density_corr));
```
Comment (sethrj) [inline src/physics/em/detail/RBEnergySampler.i.hh:57]: I think we can guarantee numerically that the second argument of the max will never be less than zero. Even if we can't, I think we should be able to allow the resulting `NaN` (from `sqrt(-epsilon)`) to propagate so that the `while` will evaluate to false and reject the sample.

File: src/physics/em/detail/RBEnergySampler.i.hh [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        dsigma       = dxsec_(gamma_energy);
+    } while (dsigma < dxsec_.maximum_value() * generate_canonical(rng));
```
Comment (sethrj) [inline src/physics/em/detail/RBEnergySampler.i.hh:59]:
```suggestion
    } while (!BernoulliDistribution(dsigma / dxsec_.maximum_value())(rng));
```
which I will add code and replace with
```c++
while (RejectionSampler(dsigma, dxsec_.maximum_value())(rng)
```

File: src/physics/em/detail/RelativisticBremDXsection.i.hh [hunk 12]
```diff
@@ -26,12 +27,11 @@ CELER_FUNCTION RelativisticBremDXsection::RelativisticBremDXsection(
...
-    real_type density_factor = material.electron_density()
-                               * this->migdal_constant();
-    density_corr_ = density_factor * ipow<2>(total_energy_);
-
-    lpm_energy_ = material.radiation_length() * this->lpm_constant();
+    real_type density_factor = material.electron_density() * migdal_constant();
+    density_corr_            = density_factor * ipow<2>(total_energy_);
 
+    lpm_energy_ = material.radiation_length()
+                  * (lpm_constant() / native_value_from(units::MevEnergy{1}));
```
Comment (sethrj) [inline src/physics/em/detail/RelativisticBremDXsection.i.hh:34]: See above units comment: this should just be:
```c++
lpm_energy_ = material.radiation_length() * value_as<MevPerCm>(lpm_constant());
```
so that `lpm_energy_` is in units of MeV which is I think what you want

Comment (sethrj) [reply src/physics/em/detail/RelativisticBremDXsection.i.hh:34]: And if `lpm_constant()` really is in native units, you should write this statement as
```c++
    lpm_energy_ = native_value_to<MevEnergy>(material.radiation_length() * lpm_constant()).value();
```

File: src/physics/em/detail/BremFinalStateHelper.hh (hunk 3 above)
Comment (whokion) [reply src/physics/em/detail/BremFinalStateHelper.hh:44]: I will take your comment below.

File: src/physics/em/detail/BremFinalStateHelper.i.hh (hunk 4 above)
Comment (whokion) [reply src/physics/em/detail/BremFinalStateHelper.i.hh:67]: Okay. Good idea.  I wanted to put the one-line momentum calculation back here to avoid the momentum argument passing multiple times (reduce memory footprint as a trade of more FLOPs, which may be a cheap?), but I guess that either of approaches should be fine.

File: src/physics/em/detail/CombinedBremInteractor.hh (hunk 5 above)
Comment (whokion) [reply src/physics/em/detail/CombinedBremInteractor.hh:36]: Done.

File: src/physics/em/detail/PhysicsConstants.hh (hunk 7 above)
Comment (whokion) [reply src/physics/em/detail/PhysicsConstants.hh:24]: Corrected.

File: src/physics/em/detail/PhysicsConstants.hh (hunk 8 above)
Comment (whokion) [reply src/physics/em/detail/PhysicsConstants.hh:46]: Okay.  Changed as suggested.

File: src/physics/em/detail/RBEnergySampler.i.hh (hunk 9 above)
Comment (whokion) [reply src/physics/em/detail/RBEnergySampler.i.hh:48]: Replaced with suggested changes for now.

File: src/physics/em/detail/RBEnergySampler.i.hh (hunk 10 above)
Comment (whokion) [reply src/physics/em/detail/RBEnergySampler.i.hh:57]: Okay. Done.

File: src/physics/em/detail/RBEnergySampler.i.hh (hunk 11 above)
Comment (whokion) [reply src/physics/em/detail/RBEnergySampler.i.hh:59]: Thanks for the suggested change - sorry to forget to use the predefined distribution.

File: src/physics/em/detail/RelativisticBremDXsection.i.hh (hunk 12 above)
Comment (whokion) [reply src/physics/em/detail/RelativisticBremDXsection.i.hh:34]: Okay, changed to value_as<MevPerCm>(lpm_constant()).

File: src/physics/em/detail/PhysicsConstants.hh (hunk 2 above)
[the remark this answers is not in the document]
Comment (sethrj) [reply src/physics/em/detail/PhysicsConstants.hh]: Technically we could directly access the particle params, since there it's not duplicated, but it would result in a slowdown for the extra indirection. I think it's better also to use the mass from the particle data rather than a hardcoded constant to have greater consistency across the codebase. The only exception to this I'm aware of is where a measurement of the electron mass from CODATA 2018 is embedded into the classical electron radius, which is then used to derive some constants. To remove *that* inconsistency we could define a function to return the "constant" as a function of the electron mass, so we could store migdal's constant etc. as `__constant__` data rather than as compile-time constants... but that seems like a lot of effort to avoid a potential inconsistency in the 7th digit (or whatever) of our constants.

**Review by sethrj — APPROVED** [review]
Looks good now. Just to confirm @whokion: you intend to have a separate MR to allow the Brems process an option to switch between the separate/combined brems model?

**Comment by whokion** [comment]
@sethrj Thanks for the review.  Yes, it would be good to merge this MR first and open a separate MR to integrate different models/options  into the process level.
